packages/astro/src/runtime/client/hmr.ts

```diff
@@ -1,21 +1,24 @@
 /// <reference types="vite/client" />
+
 if (import.meta.hot) {
-	import.meta.hot.on('vite:beforeUpdate', async (payload) => {
-		for (const file of payload.updates) {
-			if (
-				file.acceptedPath.includes('svelte&type=style') ||
-				file.acceptedPath.includes('astro&type=style')
-			) {
-				// This will only be called after the svelte component has hydrated in the browser.
-				// At this point Vite is tracking component style updates, we need to remove
-				// styles injected by Astro for the component in favor of Vite's internal HMR.
-				const injectedStyle = document.querySelector(
-					`style[data-astro-injected="${file.acceptedPath}"]`
-				);
-				if (injectedStyle) {
-					injectedStyle.parentElement?.removeChild(injectedStyle);
+	// Vite injects `<style type="text/css">` for ESM imports of styles
+	// but Astro also SSRs with `<style>` blocks. This MutationObserver
+	// removes any duplicates as soon as they are hydrated client-side.
+	const injectedStyles = getInjectedStyles();
+	const mo = new MutationObserver((records) => {
```

Comment:
overall not against this, but is there a simpler way that doesn't involve a mutation observer? In my head, I was thinking the HMR'd style-script could be responsible for removing it's duplicate `<style>` from the page as soon as it loads.

Reply:
Unfortunately not without hacking how Vite handles styles out of the box. The existing approach was to only remove the original SSR'd style when we saw an HMR update come in, but the duplicate styles exist **until** an HMR happens, and in most cases it's reasonable that an HMR update will never happen. 

This approach also allows us to get rid of `data-astro-injected="src/styles/index.css"` all over our dev output. Overall this seemed like the most resilient and unobtrusive solution.